docs/src/design/proposals/APIRedesign.md

```diff
@@ -0,0 +1,1006 @@
+recognize which constructor arguments are from dependencies, and
+which will be provided at run-time. Instead, it is the responsibility
+of whoever is introducing a component to manage these things
+separately.
+
+### Benefits
+
+* More clarity.
+
+### Detriments
+
+* Possibly results in redundant effort to manage this difference
+  (other APIs may need to be adjusted accordingly.)
+
+## Rename Views to Applications
+
+Rename (internally to the application, not necessarily in UI or
+in user guide) what are currently called `views` to `applications`.
+
+### Benefits
+
+* Easier to understand. What is currently called a "view" is,
+  in the MVC sense, a view-controller pair, usually with its own
+  internal model for view state. Calling these "applications"
+  would avoid this ambiguity/inconsistency.
+* Also provides an appropriate mindset for building these;
+  particularly, sets the expectation that you'll want to decompose
+  this "application" into smaller pieces. This nudges developers
+  in appropriate directions (in contrast to `views`, which
+  typically get implemented as templates with over-complicated
+  "controllers".)
+
+### Detriments
+
+* Developer terminology falls out of sync with what is used in
+  the user guide.
+
+## Provide Classes for Extensions
+
+As a general pattern, when introducing extension categories, provide
+classes with a standard implementation of these interfaces that
+plugin developers can `new` and register.
+
+For example, instead of declaring a type as:
+
+```json
+{
+  "types": [{
+    "key": "sometype",
+    "glyph": "X",
+    "etc": "..."
+  }]
+}
+```
+
+You would register one as:
+
+```js
+// Assume we have gotten a reference to a type registry somehow
+typeRegistry.register(new Type({
+    "key": "sometype",
+    "glyph": "X",
+    "etc": "..."
+}));
+```
+
+### Benefits
+
+* Easier to understand (less "magic").
+* Lends itself naturally to substitution of different implementations
+  of the same interface.
+* Allows for run-time decisions about exactly what gets registered.
+
+### Detriments
+
+* Adds some modest boilerplate.
+* Provides more opportunity to "do it wrong."
+
+## Normalize naming conventions
+
+Adopt and obey the following naming conventions for AMD modules
+(and for injectable dependencies, which may end up being modules):
+
+* Use `UpperCamelCase` for classes.
+* Use `lowerCase` names for instances.
+    * Use `someNoun` for object instances which implement some
+      interface. The noun should match the implemented interface,
+      when applicable.
+    * `useSomeVerb` for functions.
+* Use `ALL_CAPS_WITH_UNDERSCORES` for other values, including
+  "struct-like" objects (that is, where the object conceptually
+  contains properties rather than methods.)
+
+### Benefits
+
+* Once familiar with the conventions, easier to understand what
+  individual modules are.
+
+### Detriments
+
+* A little bit inflexible.
+
+## Expose no third-party APIs
```

Comment:
Pulling related comments from #229:

**[@larkin wrote](https://github.com/nasa/openmctweb/issues/229#issuecomment-153438408)**:
> @VWoeltjen whoa, that's something I hadn't had time to process... "expose no third party APIs?"
> 
> I'd prefer we reuse and expose APIs that people are familiar with-- for one, it requires less documentation and maintenance on our part. two, it stops us from becoming a monolithic framework that is a career commitment to learn, and three, it drastically lowers the barrier to entry.
> 
> Many people have used lodash/underscore. Many people have used angular. Many people have used requirejs. Having to learn a new thing that is similar, but different than all three? That's hard.

**[@akhenry wrote](https://github.com/nasa/openmctweb/issues/229#issuecomment-153443890)**:
> I tend to agree with @larkin. Arguably it complicates our own API and actually increases the learning curve if we end up having to replicate or wrap functionality already provided by popular third-party APIs that people are familiar with.
> 
> I completely understand the motivation though. The world of web development is horribly fickle and frameworks that are ubiquitous now won't be in a few years. prototype.js anyone? Even the once mighty jquery has fallen out of favour.
> 
> I think we should limit the third party APIs we expose to one or two, but I worry it might be counterproductive to completely hide them.
> 
> One option might be to provide lightweight facades to third party APIs ala Angular with its quasi-jQuery interface, but that feels a bit blergh.

**[@vwoeltjen wrote](https://github.com/nasa/openmctweb/issues/229#issuecomment-153446336)**:
> @larkin - I think that's the wrong conclusion to reach there. There's nothing stopping you from using lodash/require/etc. directly; we're compatible with other libraries and should try to make that compatibility simpler. In general, I don't think we should be re-exposing functionality that can be directly obtained from other more general-purpose libraries.
> 
> The idea that exposing third-party libraries reduces the learning curve has been directly contradicted by feedback we've gotten. We've regularly heard that "to use this I have to learn Open MCT Web, Angular, and Require." That's fine internally, but really it would be better if it was "to use this I have to learn Open MCT Web, and I can use it with the APIs I already know."
> 
> I think this is particularly important for a 1.0.0 release; if we're exposing third-party APIs at that point, we're committed to those dependencies until 2.0.0. Wrapping the things we want/need to expose narrows that commitment.
> 
> Angular's a bit of a weird case because it provides quite a useful services library and templating syntax. My solution there would be (a) use that internally, or wrap it if we want to expose something similar, and (b) provide a non-platform Angular plug-in that does expose these things.

**[@larkin wrote](https://github.com/nasa/openmctweb/issues/229#issuecomment-153453035)**:
> Whether you're using require, browserify or plain script tags, the concept of a dependency loader is something you should be familiar with in every language. People not interested in gaining a cursory understanding of these tools are really saying "I don't want to learn anything about web technologies". Do we really want to cater to these people?
> 
> Lodash is a way of simplifying common functional patterns and reducing code. Don't think we need to expose it, but don't think we need to recreate it. If we use it in the platform, then it's there if people want it, but if they don't import it, they don't have to use it.
> 
> If you look at Angular tutorials, they look nothing like the code we write. Angular is itself a monolith that tries to promote good programming practices, it handles many concerns that we have reimplemented separately. So telling people it's angular, well, that isn't very helpful. If it actually looked like angular, that would have more value.
> 
> When people say "I have to learn X, Y, and Z to use this" and then claim that they'd prefer to say "I only have to learn Z," well, if Z included all of X, Y, and Z, then they'd still have to learn it all. I don't think it's because it's too many tools; I think the problem is it's not clear what the boundaries are between X, Y, and Z.[1]
> 
> As much as we call openmctweb a platform, I would better define it as a set of tools, common interfaces, and design patterns that is intended to be easily extensible and promote good development practices.
> 
> [1]: Haven't done web development before? You're already in over your head: you've also got to learn A, B, and C too.